python/dries_json_reader.py

```diff
@@ -48,6 +48,75 @@ def cells_json(data):
     return cell_dict
 
 
+def cell_sets_json(data):
+    '''
+    >>> data = {
+    ...     "cell_1": {
+    ...         "locations": [1632.02, -1305.7],
+    ...         "factors": {
+    ...             "pleiden_clus": [3],
+    ...             "kmeans": [8]
+    ...         },
+    ...         "mappings": {
+    ...             "tsne": [-11.0776, 6.0311],
+    ...             "umap": [-6.858, 15.7691]
+    ...         }
+    ...     }
+    ... }
+    >>> cell_sets = cell_sets_json(data)
+    >>> list(cell_sets.keys())
+    ['version', 'datatype', 'tree']
+    >>> cell_sets['datatype']
+    'cell'
+    >>> len(cell_sets['tree'])
+    2
+    >>> sorted([ n['name'] for n in cell_sets['tree'] ])
+    ['Leiden Clustering', 'k-means Clustering']
+    '''
+    clustering_dict = {
+        'pleiden_clus': {},
+        'kmeans': {}
+    }
+    nice_names = {
+        'pleiden_clus': 'Leiden Clustering',
+        'kmeans': 'k-means Clustering'
+    }
+    for cell_id in data.keys():
+        factors = data[cell_id]['factors']
+        factors_dict = {
+            'pleiden_clus': 'Cluster {}'.format(factors['pleiden_clus'][0]),
+            'kmeans': 'Cluster {}'.format(factors['kmeans'][0])
+        }
+        # For each cluster assignment, append this cell ID to the
+        # appropriate clustering_dict list.
+        for factor_type, factor_cluster in factors_dict.items():
+            if factor_cluster in clustering_dict[factor_type]:
+                clustering_dict[factor_type][factor_cluster].append(cell_id)
+            else:
+                clustering_dict[factor_type][factor_cluster] = [cell_id]
+
+    # Construct the tree
+    cell_sets = {
+        'version': '0.1.2',
+        'datatype': 'cell',
+        'tree': []
+    }
+
+    for factor_type, factor_clusters in clustering_dict.items():
+        factor_type_children = []
+        for cluster_name in sorted(factor_clusters.keys()):
```

Comment:
Use comprehension instead?

Reply:
Please see previous comment about original comprehension implementation being more difficult to understand.